pyanaconda/modules/timezone/initialization.py

```diff
@@ -0,0 +1,75 @@
+#
+# Copyright (C) 2022  Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+import time
+from pyanaconda.core.configuration.anaconda import conf
+from pyanaconda.core.constants import NETWORK_CONNECTION_TIMEOUT
+from pyanaconda.modules.common.constants.services import NETWORK
+from pyanaconda.modules.common.task import Task
+from pyanaconda.modules.common.util import is_module_available
+from pyanaconda.modules.timezone.geoloc import provider_id_to_class
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class GeolocationTask(Task):
+    """Run geolocation"""
+    @property
+    def name(self):
+        return "Geolocate the system"
+
+    def run(self):
+        provider_class = provider_id_to_class(conf.ui.geolocation_provider)
+        provider = provider_class()
+
+        log.info("Geoloc: starting lookup using provider: %s", provider.name)
+        start_time = time.time()
+
+        if not self._wait_for_network():
+            log.error("Geoloc: no network connection")
+            return None
+
+        result = provider.locate()
+        log.info("Geoloc: lookup finished in %1.1f seconds", time.time() - start_time)
+
+        if not result:
+            log.info("Geoloc: no result")
+            return None
+
+        log.info("Geoloc: got result: %s", result)
+        return result
+
+    def _wait_for_network(self, timeout=NETWORK_CONNECTION_TIMEOUT):
+        """Wait until network is available, or time runs out
+
+        :param float timeout: how long shall we try waiting
+        :return bool: is there network connectivity
+        """
+        if not is_module_available(NETWORK):
+            return False
+
+        network = NETWORK.get_proxy()
+
+        waited = 0
+        while waited < timeout:
+            waited = waited + 0.3
+            time.sleep(0.3)
+            if network.Connected:
```

Comment:
You can wait for the `PropertiesChanged` signal instead of the active waiting if you want to.

Reply:
I don't see how to realize a timeout in such case. That might not be a problem, though.

Reply:
You will need a synchronization object from the `threading` library. This object can block the task thread until a timeout or until it is released by the `PropertiesChanged` callback (the callback runs in the main thread). See: https://docs.python.org/3/library/threading.html

Reply:
After coming back to this after a few days, I realized - one benefit of using a thread (Task) should be to write the code synchronously. So if possible, I'd leave this out, or for later.